gsplat/cuda/csrc/projection.cu

```diff
@@ -1582,3 +1582,397 @@ fully_fused_projection_packed_bwd_tensor(
     }
     return std::make_tuple(v_means, v_covars, v_quats, v_scales, v_viewmats);
 }
+
+//====== 2DGS ======//
+__global__ void
+fully_fused_projection_fwd_2dgs_kernel(const uint32_t C, const uint32_t N,
+                                    const float *__restrict__ means,    // [N, 3]
+                                    const float *__restrict__ quats,    // [N, 4]
+                                    const float *__restrict__ scales,   // [N, 3]
+                                    const float *__restrict__ viewmats, // [C, 4, 4]
+                                    const float *__restrict__ Ks,       // [C, 3, 3]
+                                    const int32_t image_width, const int32_t image_height,
+                                    const float eps2d, const float near_plane,
+                                    const float far_plane, const float radius_clip,
+                                    // outputs
+                                    int32_t *__restrict__ radii,            // [C, N]
+                                    float *__restrict__ means2d,            // [C, N, 2]
+                                    float *__restrict__ depths,             // [C, N]
+                                    float *__restrict__ ray_transformations // [C, N, 3, 3]
+) {
+    // parallelize over C * N
+    uint32_t idx = cg::this_grid().thread_rank();
+    if (idx >= C * N) {
+        return;
+    }
+    const uint32_t cid = idx / N; // camera id
+    const uint32_t gid = idx % N; // gaussian id
+    
+    // shift points to the current camera and gaussian
+    means += gid * 3;
+    viewmats += cid * 16;
+    Ks += cid * 9;
+
+    // glm is column-major but input is row-major
+    glm::mat3 R = glm::mat3(viewmats[0], viewmats[4], viewmats[8], // 1st column
+                            viewmats[1], viewmats[5], viewmats[9], // 2nd column
+                            viewmats[2], viewmats[6], viewmats[10] // 3rd column
+    );
+    glm::vec3 t = glm::vec3(viewmats[3], viewmats[7], viewmats[11]);
+
+    // transform Gaussian center to camera space
+    glm::vec3 mean_c;
+    pos_world_to_cam(R, t, glm::make_vec3(means), mean_c);
+    if (mean_c.z < near_plane || man_c.z > far_plane) {
+        radii[idx] = 0;
+        return;
+    }
+
+    // build H and transform from world space to camera space
+    glm::mat3 RS = quat_to_rotmat(quat) * scale_to_mat(scale, 1.0f);
+    glm::mat3 RS_camera = R * RS;
+    glm::mat3 WH = glm::mat3(RS_camera[0], RS_camera[1], mean_c);
+
+    glm::mat3 inverse_intrinsic = glm::mat3(
```

Comment:
hmmm, the ellipse aabb in the original paper uses the projection matrix which transforms into clipping space, I'm not sure if we can do this and directly take the same computation for ellipse aabb? @FantasticOven2 @liruilong940607 am I misunderstanding this?

Reply:
I think using ray-splat intersection to compute AABB is more accurate than using the perspective transformation for ellipse aabb in the original paper as mentioned [here](https://github.com/TimSong412/2D-surfel-gaussian/issues/13#issue-2324615894); the author also worked out the math [here](https://github.com/hbb1/diff-surfel-rasterization/issues/8#issuecomment-2138069016)